src/zelos/emulator/base.py

```diff
@@ -186,37 +656,94 @@ def unpack(
         Unpacks an integer from a byte format. Defaults to the
         current architecture bytes and endianness.
         """
-        endian_char, bit_char, bit_mask = self._pack_format(
-            bytes, little_endian, signed
-        )
-        return struct.unpack(endian_char + bit_char, x)[0]
-
-    def _pack_format(self, bytes: int, little_endian: bool, signed: bool):
-        """
-        Generates the format for the struct.pack and unpack functions.
-        Defaults to the current architecture bytes and endianness
-        """
         if bytes is None:
             bytes = self.bytes
         if little_endian is None:
-            little_endian = self.state.endianness == "little"
-        bits = bytes * 8
-
-        bit_char = {8: "B", 16: "H", 32: "I", 64: "Q"}[bits]
-        bit_mask = 2 ** bits - 1
-
-        if signed:
-            bit_char = bit_char.lower()
-        endian_char = "<" if little_endian else ">"
-        return endian_char, bit_char, bit_mask
+            little_endian = self.state.endianness[0] == "l"
+        fmt = None
+        if little_endian:
+            if signed:
+                fmt = self._pack_fmt_signed_little[bytes]
+            else:
+                fmt = self._pack_fmt_little[bytes]
+        else:
+            if signed:
+                fmt = self._pack_fmt_signed_big[bytes]
+            else:
+                fmt = self._pack_fmt_big[bytes]
+        return unpack(fmt, x)[0]
+
+    def _mem_map_region(self, mr: MemoryRegion):
+        self._uc.mem_map_ptr(mr.address, mr.size, mr.prot, mr.host_address)
+        self._page_table.add(mr)
+        self._regions[mr.address] = mr
+
+    def _mem_unmap_region(self, mr: MemoryRegion):
+        self._uc.mem_unmap(mr.address, mr.size)
+        self._page_table.remove(mr)
+        del self._regions[mr.address]
+
+    def _mem_area_overlaps(self, address: int, size: int) -> bool:
+        end = address + size - 1
+        for mr in self.mem_regions():
+            if address >= mr.address and address <= mr.end:
+                return True
+            if end >= mr.address and end <= mr.end:
+                return True
+            if address < mr.address and end > mr.end:
+                return True
+
+    def _mem_area_mapped(self, address: int, size: int) -> bool:
+        count = 0
+        while count < size:
+            mr = self.mem_region(address)
+            if mr is None:
+                break
+            length = min(size - count, mr.end + 1 - address)
+            count += length
+            address += length
+        return count == size
+
+    def _split_region(
+        self,
+        mr: MemoryRegion,
+        address: int,
+        size: int,
+        do_delete: bool = False,
+    ):
+        chunk_end = address + size
+        if (address <= mr.address and chunk_end > mr.end) or size == 0:
+            return
+        if address > mr.end or chunk_end <= mr.address:
+            raise OutOfMemoryException()
```

Comment:
Is this an OutOfMemoryException? Seems like this address is just not within the MR?